Add autoscheduler support to tvmc

giuseros opened the pull request:
- Add an autoschedule module to tvmc
- Extract common tuning option between autotuner and autoscheduler
- Add testing

giuseros commented:
cc @leandron @merrymercy

comaniac commented on python/tvm/driver/tvmc/autoscheduler.py:
```diff
@@ -0,0 +1,212 @@
+        logger.debug("Default --min-repeat-ms for this target is %s", min_repeat_ms)
+
+    if args.rpc_tracker:
+
```
Remove this line.

comaniac commented on python/tvm/driver/tvmc/autoscheduler.py:
```diff
@@ -0,0 +1,212 @@
+        num_measure_trials=args.trials,
+        measure_callbacks=[auto_scheduler.RecordToFile(args.output)],
+        runner=runner,
+        builder="local",
```
No need to provide the default value.

comaniac commented on python/tvm/driver/tvmc/autoscheduler.py:
```diff
@@ -0,0 +1,212 @@
+    )
+
+
+def get_tuning_tasks(
```
This function looks unnecessary, and it's a bit weird to perform layout conversion in a function called "get_tuning_tasks".

giuseros replied:
File has been deleted

comaniac commented on python/tvm/driver/tvmc/autoscheduler.py:
```diff
@@ -0,0 +1,212 @@
+    Parameters
+    ----------
+    tasks : list
+        A list of autotvm.Tasks to tune.
```
```suggestion
        A list of auto_scheduler.SearchTask to tune.
```

giuseros replied:
File has been deleted

comaniac commented on python/tvm/driver/tvmc/autoscheduler.py:
```diff
@@ -0,0 +1,212 @@
+        A list of autotvm.Tasks to tune.
+    task_weights : list
+        The weight (i.e. the number of appearance) of extracted tasks
+    tuning_records : str, optional
+        The json file used to preload the autoscheduler
+    tuning_options:
+        The options of tuning
```
Wrong order.

giuseros replied:
File has been deleted but I applied this change

comaniac commented on python/tvm/driver/tvmc/compiler.py:
```diff
@@ -82,6 +82,11 @@ def add_compile_parser(subparsers):
         help="path to an auto-tuning log file by AutoTVM. If not presented, "
         "the fallback/tophub configs will be used",
     )
+    parser.add_argument(
+        "--use-autoscheduler",
+        action="store_true",
+        help="use the autoscheduler to generate the compute schedules",
+    )
```
You actually don't need users to tell you if auto_scheduler is used. Since AutoTVM and AutoScheduler have different log formats, you can simply use the following logic to know which one you should use:

```python
def is_auto_scheduler_log(log_file):
    try:
        auto_scheduler.load_records(log_file)
    except Exception:
        return False
    return True
```

giuseros replied:
Very nice suggestion, thanks!

comaniac commented on tests/python/driver/tvmc/test_autoscheduler.py:
```diff
@@ -0,0 +1,101 @@
+    return (tasks, weights)
+
+
+def _autoscheduler_test_helper(
```
I'm not sure if we need to actually run auto_scheduler in this test. It seems to me that this helper is used by the following 4 test cases. However, TVMC should not be in charge of testing "tasks", "tuning records", "early stopping" and "no tuning records", especially those tests don't even use TVMC commands.

giuseros replied:
I removed this file. But now the autoscheduler is untested.

comaniac commented on python/tvm/driver/tvmc/autoscheduler.py:
```diff
@@ -0,0 +1,212 @@
+@register_parser
+def add_autoscheduler_parser(subparsers):
+    """ Include parser for 'autoschedule' subcommand """
+    parser = subparsers.add_parser("autoschedule", help="auto-schedule a model")
```
`--enable-auto-scheduler` looks good to me, or we could use `--use-auto-scheduler`, which aligns the configuration name used in the PassContext.

giuseros replied:
I went for `--enable-autoscheduler` (without the dash). This means that I can get rid of a lot of the `common` code. Very good!

giuseros commented on python/tvm/driver/tvmc/autoscheduler.py:
```diff
@@ -0,0 +1,212 @@
+        "--vector-unit-bytes",
+        default=16,
+        help="the width of vector units in bytes",
+    )
```
I used two groups actually, to differentiate between autotuner and autoschedule options

giuseros commented on python/tvm/driver/tvmc/common.py:
```diff
@@ -36,6 +36,93 @@ class TVMCException(Exception):
     """TVMC Exception"""
 
 
+def add_tuning_options(parser):
```
Got rid of this

comaniac replied:
Based on the latest refactor you don't need this function anymore?

giuseros commented on python/tvm/driver/tvmc/autoscheduler.py:
```diff
@@ -0,0 +1,212 @@
+        min_repeat_ms = args.min_repeat_ms
+    else:
+        min_repeat_ms = 0 if target.keys[0] == "cpu" else 1000
+        logger.debug("Default --min-repeat-ms for this target is %s", min_repeat_ms)
```
File has been deleted

comaniac commented on python/tvm/driver/tvmc/autotuner.py:
```diff
@@ -196,28 +237,55 @@ def drive_tune(args):
 
 
-def get_tuning_tasks(mod, params, target, target_host=None, alter_layout=None):
+def autotuner_get_tuning_tasks(mod, params, target, target_host=None, alter_layout=None):
```
```suggestion
def autotvm_get_tuning_tasks(mod, params, target, target_host=None, alter_layout=None):
```

comaniac commented on python/tvm/driver/tvmc/autotuner.py:
```diff
@@ -253,6 +321,72 @@ def get_tuning_tasks(mod, params, target, target_host=None, alter_layout=None):
+def autoscheduler_get_tuning_tasks(
+    mod, params, target, target_host=None, alter_layout=None, hardware_params=None
+):
+    """Get the tuning tasks for a given relay module.
```
```suggestion
    """Get the auto-scheduler tuning tasks for a given relay module.
```
Ditto to the docstring in `autotuner_get_tuning_tasks`.

comaniac commented on python/tvm/driver/tvmc/autotuner.py:
```diff
@@ -253,6 +320,72 @@ def get_tuning_tasks(mod, params, target, target_host=None, alter_layout=None):
+        The compilation target.
+    target_host : str, optional
+        The compilation target for the host.
+    alter_layout : str, optional
+        The layout to convert the graph to. Note, the convert layout
+        pass doesn't currently guarantee the whole of the graph will
+        be converted to the chosen layout.
```
Wrong doc.

comaniac commented on python/tvm/driver/tvmc/autotuner.py:
```diff
@@ -253,6 +320,72 @@ def get_tuning_tasks(mod, params, target, target_host=None, alter_layout=None):
+        A list of auto_scheduler.SearchTask to tune.
+    task_weights : list
+        The weight (i.e. the number of appearance) of extracted tasks
+    tuning_options:
```
```suggestion
    tuning_options: dict
```

giuseros commented:
Hi @comaniac , @leandron , 
Thanks for your comments, now the code is way more sound. I thought of summing up here some issues left:
* It's now hard to make autotuner vs autoscheduler options mutually exclusive. So if I specify `--tuner` with `--enable-autoscheduling`, the tuner I selected won't be considered. This is because we are using default values, so a tuner will be always defined. 
* Extracting tasks. @comaniac , you flagged the function, but this is actually how we do for the autotuner. I guess this is because we want to unit-test the function
* Testing: I removed the `test_autoscheduler.py` test. When I wrote it, I was inspired by the `test_autotuner.py`. Can we agree on a set of tests we want to execute? @comaniac , I agree we should not test the autotuner, but only the user interface. But I fear to test the user interface, we will need to test a bit the functionality as well. What do you guys think?

comaniac commented on python/tvm/driver/tvmc/compiler.py:
```diff
@@ -182,10 +182,30 @@ def compile_model(
-                graph_module = relay.build(mod, tvm_target, params=params, target_host=target_host)
+
+        use_autoscheduler = True
+
```
Remove this line.

comaniac reviewed (changes requested):
It's much clean now and overall LGTM. Thanks!
The last batch of comments are mostly about refactoring.

comaniac commented:
> Hi @comaniac , @leandron ,
> Thanks for your comments, now the code is way more sound. I thought of summing up here some issues left:
> 
> * It's now hard to make autotuner vs autoscheduler options mutually exclusive. So if I specify `--tuner` with `--enable-autoscheduling`, the tuner I selected won't be considered. This is because we are using default values, so a tuner will be always defined.

I'm ok with this for now.

> * Extracting tasks. @comaniac , you flagged the function, but this is actually how we do for the autotuner. I guess this is because we want to unit-test the function

No worries. It is no longer an issue based on the latest implementation. It is reasonable to have separate functions for two frameworks launched by the same CLI command.

> * Testing: I removed the `test_autoscheduler.py` test. When I wrote it, I was inspired by the `test_autotuner.py`. Can we agree on a set of tests we want to execute? @comaniac , I agree we should not test the autotuner, but only the user interface. But I fear to test the user interface, we will need to test a bit the functionality as well. What do you guys think?

You're right. It's always hard to keep the proper unit test scope of TVMC. We should have a thread to discuss all of them and make a agreement. cc @leandron

giuseros commented on python/tvm/driver/tvmc/autotuner.py:
```diff
@@ -133,6 +127,45 @@ def add_tune_parser(subparsers):
+    auto_tuning_group = parser.add_argument_group(
+        "Autotuning options",
+        "Autotuning options, used when the autoscheduler is not enabled",
+    )
```
Since this is also for me to familiarize with the Autoscheduler, I would prefer to add here anything that you and @merrymercy think relevant in terms of Autoscheduler options.

giuseros commented:
Hi @comaniac , @leandron , 
I updated the PR as follows:
* I added two more options (--log-estimated-latency and --include-simple-tasks)
* I added the tests back (as they were before). 
Please, let me know if those tests are ok (or if I need to add more or less), and if there is any other autoscheduler relevant option I need to consider. 

Thanks,

comaniac commented on python/tvm/driver/tvmc/autotuner.py:
```diff
@@ -133,6 +127,85 @@ def add_tune_parser(subparsers):
+        help="Whether to log the estimated latency to the file after tuning a task",
+        action="store_true",
+    )
+    auto_tuning_group = parser.add_argument_group(
```
Replace all `auto_tuning` with `autotvm` to be consistent as the underlying TVMC implementations.

comaniac reviewed (approved):
Overall LGTM. Thanks. Just a naming nit.

insop commented on python/tvm/driver/tvmc/autotuner.py:
```diff
@@ -133,6 +127,85 @@ def add_tune_parser(subparsers):
+    )
+    auto_scheduler_group.add_argument(
+        "--log-estimated-latency",
+        help="Whether to log the estimated latency to the file after tuning a task",
```
Nit, all other help text starts with low case, except the above two texts.

comaniac commented:
Thanks @giuseros @leandron @insop.

merrymercy commented:
For x86 cpu, it is recommended to use `enable_cpu_cache_flush=True` in `LocalRunner`, as shown in the tutorial.
https://github.com/apache/tvm/blob/54c995dbf7c96c1184c2baf64de87bc9566fe33a/tutorials/auto_scheduler/tune_network_x86.py#L187
This can make the measurement faster and more accurate. We should add this to tvmc.

giuseros commented:
Hi @merrymercy ,
Thanks for the comment! Sure, I will open a new PR to add this in
